### PR TITLE
Inline equations linebreak at appropriate places

### PR DESCRIPTION
**Pull request by damaxwell**
Currently, Typst treats an inline equation as a monolithic item.  No line breaks are possible in the middle.

This PR implements something close to TeX's strategy for allowing line breaks.  A break is allowed:
* After a binary operator or a relation,
* unless the operator or relation is enclosed in a left-right group (or in a more complicated thing like an exponent or a radicand),
* except that a relation followed by a relation doesn't line break.
* Notwithstanding the above, inline equations with more than one line never line break.

This ended up being somewhat trickier than I anticipated.  The paragraph rendering algorithm delayed rendering of equations and boxes until late, but needed to know how many items in the paragraph these objects will become early on.  That's easy if the answer is always "1", but for equations the answer now depends on layout.  So layout was migrated sooner in the algorithm.

Additionally, when determining whether a line break is allowed after all of the initial math layout is done, you need to know if binary ops and relations are sitting inside of LR groups.  But not enough information was left behind to determine this.  So an additional layer of tags was introduced so that beginnings/ends of LR groups could be identified after the fact.

Some notes/questions:
1) Meta information for the whole inline equation is applied to all of the resulting (non-space) fragments. I need confirmation that that was the right choice.
2) The `Layout` trait takes a `&mut Vt` as an argument.  I think that means, in principle, that there can be ordering effects if children are laid out in different orders.  So when I moved the equation rendering early, I also moved the box rendering early.  But I noticed that layout of fractional-spaced boxes was, in fact, already delayed until the very end of paragraph layout, just before completion. Are ordering effects a non-issue? I'm not exactly sure what in the `Vt` might be altered during layout.  At any rate, this PR lays out paragraph children in the same order as was done previously (preserving any bugs, if they were present).
3) In TeX, something like `$10^3$cm` does not allow a line break between the formula and the text.  By contrast, Typst allows line breaks here.  One could implement this, I think, by changing the way formula fragments are represented in the paragraph (as a non-breaking space U+00A0 instead of an object replacement U+FFFC, perhaps). But this is best tackled, if desired, in a separate PR.
4) A couple of test source files were edited so that newly legal line breaks did not interfere with the spirit of the test.

## Example
```
#set page(width:120pt,margin:10pt,height:auto)
#let hrule(x) = {box(line(length:x))}
#hrule(50pt)$e^(pi i)+1 = 0$\
#hrule(60pt)$e^(pi i)+1 = 0$\
#hrule(80pt)$e^(pi i)+1 = 0$\
```
### Before
image

### After
image

### CODE REVIEW

**Comment by laurmaedje** [comment]
> Meta information for the whole inline equation is applied to all of the resulting (non-space) fragments. I need confirmation that that was the right choice.

Should be fine. The introspector will deduplicate it.

> The Layout trait takes a &mut Vt as an argument. I think that means, in principle, that there can be ordering effects if children are laid out in different orders. So when I moved the equation rendering early, I also moved the box rendering early. But I noticed that layout of fractional-spaced boxes was, in fact, already delayed until the very end of paragraph layout, just before completion. Are ordering effects a non-issue? I'm not exactly sure what in the Vt might be altered during layout. At any rate, this PR lays out paragraph children in the same order as was done previously (preserving any bugs, if they were present).

The ordering isn't determined by the order in which things are layouted, but by the order in which they end up in the frames. The important thing is thus just that the order in which things are layouted _is the same across multiple layout passes._ This is because the `&mut Vt` is the source of stable identifiers (`Location`s).

> In TeX, something like $10^3$cm does not allow a line break between the formula and the text. By contrast, Typst allows line breaks here. One could implement this, I think, by changing the way formula fragments are represented in the paragraph (as a non-breaking space U+00A0 instead of an object replacement U+FFFC, perhaps). But this is best tackled, if desired, in a separate PR.

You are correct that this would require changing the way formula fragments are representation. I'm fine with deferring this, it seems unrelated.

> A couple of test source files were edited so that newly legal line breaks did not interfere with the spirit of the test.

Sounds good!

File: crates/typst-library/src/layout/par.rs [hunk 1]
```diff
@@ -313,23 +314,24 @@ impl<'a> Preparation<'a> {
 
             cursor += len;
         }
-
```
Comment (laurmaedje) [inline crates/typst-library/src/layout/par.rs:316]: I'd like to keep this line break :)

File: crates/typst-library/src/layout/par.rs [hunk 2]
```diff
@@ -340,8 +342,19 @@ impl Segment<'_> {
...
+            Self::Equation(_, ref par_items) => {
+                let mut len = 0;
+                for item in par_items {
+                    let c = match item {
+                        MathParItem::Space(_) => SPACING_REPLACE,
+                        MathParItem::Frame(_) => OBJ_REPLACE,
+                    };
+                    len += c.len_utf8();
+                }
+                len
```
Comment (laurmaedje) [inline crates/typst-library/src/layout/par.rs:356]: this could be nicely written in iterator style

**Review by laurmaedje — COMMENTED** [review]
The paragraph changes look good. I really don't like the GroupRole thing though and specifically that every fragment is now a tuple. Why have the rule with the `lr` groups in the first place? It seems to me that TeX doesn't do that either:

image

EDIT: I have to correct myself. I have not added `\left` and `\right` ... however, lr will be much more prevalent in Typst than in `\left .. \right` in TeX, so that's something to keep in mind.

**Comment by damaxwell** [comment]
The `GroupRole` thing is very much unappealing, I totally agree.

It's there because TeX's left/right groups don't break (see below) and I interpreted `lr` as the equivalent of a left/right group, which it largely is.  They do happen more frequently in Typst than in LaTeX  because of automatic delimiter matching. (Oh, I see; you've updated your comment; you know all this)

We can drop the `GroupRole` thing and allow breaking in left/right groups.  

Alternatively we can rejigger `MathFragment` so that it's a single object that has two fields: one is a current `MathFragment` enum and one is a `GroupRole`.   At least that way there are fewer `GroupRole`s announcing themselves in the code.  Probably that would have been the better thing to have done in the first place.

----

The LaTeX code:
```
\documentclass[12pt]{article}
\usepackage[papersize={200pt, 100pt},left=20pt, right=20pt]{geometry}
\begin{document}
\noindent fill fill fill fill fill fill fill $\left(a+b+c\right)$
\end{document}
```
yields
image

Add one more `fill` and you get
image

And one more and you get
image

**Comment by laurmaedje** [comment]
To be honest, I think breaking in LR is not that bad. It's still only after relations.

**Comment by damaxwell** [comment]
I'll update to do that, and will address the other minor comments.

**Comment by CGMossa** [comment]
How's it going with this @damaxwell?
It is really a crucial feature, and I'm very happy that someone too it up!

**Comment by laurmaedje** [comment]
Closing due to inactivity.

**Comment by davidleejy** [comment]
Can't wait til this feature drops in the new Typst release 😄

**Pull request closed without merging**